### PR TITLE
SQ-288/Scroll to next event

### PR DESCRIPTION
**Pull request by danybony**
## Problem

We want to automatically scroll to the current/next event slot when opening app.
This is related to https://github.com/squanchy-dev/squanchy-android/issues/288

## Solution

* When the schedule is updated, select the right tab and scroll (not smoothly) to the next slot automatically (only if we are in the days of the conference).
* We also added the option for the user to tap the day tab to automatically scroll (smoothly) to the next slot, in case the app was opened without a refresh / the user navigated to other times
* To ease testing, we used the `CurrentTimeModule` to specify the current time: in production will use the system clock while in debug it's using an hardcoded date. That date is now to the Droidcon London 2017, in a future PR we'll update it to be selectable from the debug activity.

**Note:**
Me and @dextorer are in disagreement regarding one aspect of the implementation: which slot to jump on. The selected option here is to jump to the *next* slot compared to the current one, while the alternative option is to jump to the *current* slot.
Feedbacks are highly appreciated here.

### Test(s) added 

Nope, manual testing 👍 

### Paired with 

@dextorer

### CODE REVIEW

**Comment by rock3r** [comment]
Re: which slot to jump to, it would seem more sensible to jump to the current slot, just like a calendar opens by default to the current month/week/day/hour

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 1]
```diff
@@ -120,25 +124,51 @@ class SchedulePageView @JvmOverloads constructor(
     }
 
     fun updateWith(schedule: Schedule, onEventClicked: (Event) -> Unit) {
-        viewPagerAdapter.updateWith(schedule.pages, onEventClicked)
+        val initialEventForPage = schedule.pages.map(::findNextEventForPage).toTypedArray()
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: I was puzzling over this line, the name got me very confused as to why there was a `toTypedArray()` at the end. Why not calling it `currentEventByPage` (assuming you make it select the _current_ event)?

I also don't understand why you need `toTypedArray()`. Why not leaving it as a `List<Event?>`

File: app/src/main/java/net/squanchy/support/system/CurrentTime.kt [hunk 2]
```diff
@@ -1,10 +1,11 @@
 package net.squanchy.support.system
 
+import net.squanchy.BuildConfig
 import org.joda.time.LocalDateTime
 
 class CurrentTime {
 
-    fun currentTimestamp(): Long = System.currentTimeMillis()
+    private val mockTime = LocalDateTime.parse("2017-10-27T12:00:00.000")
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: ❌ why is a mock value in production code?

File: app/src/main/java/net/squanchy/support/system/CurrentTime.kt [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
 
-    fun currentLocalDateTime(): LocalDateTime = LocalDateTime.now()
+    fun currentLocalDateTime(): LocalDateTime = if (BuildConfig.DEBUG) mockTime else LocalDateTime.now()
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: ❌ this switch should be done in Dagger, not here!

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 4]
```diff
@@ -120,25 +124,51 @@ class SchedulePageView @JvmOverloads constructor(
...
+
+        val todayPageIndex = findTodayIndexOrDefault(schedule.pages)
+        viewpager.setCurrentItem(todayPageIndex, false)
+
+        tabstrip.addOnTabSelectedListener(ScrollingOnTabSelectedListener(schedule, viewPagerAdapter))
         progressbar.visibility = View.GONE
     }
 
+    private fun findTodayIndexOrDefault(pages: List<SchedulePage>): Int {
+        val now = currentTime.currentLocalDateTime()
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: ⚠️ I am 99.9999% sure that this won't work correctly across timezones. `LocalDateTime` does not take those into account; if you're running the app in a TZ on which the day is different than the current day at the conference venue (to which all event `LocalDateTime`s refer to), this will break and either select the wrong day, or not find a day to select and fall in the default case.

Happy to be proved wrong, but given there is no tests around it, I don't know if this edge case has been covered.

**Comment by quiro91** [comment]
Agree with the current slot, I understand that the user might want to see the next slot but at that point it's going to be easy to scroll down to see the next one

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 5]
```diff
@@ -120,25 +124,51 @@ class SchedulePageView @JvmOverloads constructor(
...
+        return pages.firstOrNull { it.date.toLocalDate().isEqual(now.toLocalDate()) }
+            ?.let(pages::indexOf) ?: 0
+    }
+
+    private fun findNextEventForPage(page: SchedulePage) =
+        page
+            .events
+            .firstOrNull {
+                it.startTime.toDateTime(it.timeZone).isAfter(currentTime.currentLocalDateTime().toDateTime(it.timeZone))
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: This line should be chopped down, it's hard to parse

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 6]
```diff
@@ -34,6 +36,7 @@ class SchedulePageView @JvmOverloads constructor(
     private val service: ScheduleService
     private val navigate: Navigator
     private val analytics: Analytics
+    private val currentTime: CurrentTime
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt:38]: I'm wondering, why isn't the current page and event data calculated in the data layer? It really doesn't seem to belong to the UI layer. It should likely be part of `Schedule`, in form of `currentPageIndex` and `currentEventIndex`

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt [hunk 7]
```diff
@@ -52,3 +57,19 @@ class ScheduleDayPageView @JvmOverloads constructor(
         override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) = oldEvents[oldItemPosition] == newEvents[newItemPosition]
     }
 }
+
+private class SnappingLinearLayoutManager(context: Context) : LinearLayoutManager(context) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: I'd put this and `TopSnappedSmoothScroller` in their own file

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt [hunk 8]
```diff
@@ -52,3 +57,19 @@ class ScheduleDayPageView @JvmOverloads constructor(
...
+
+    override fun smoothScrollToPosition(recyclerView: RecyclerView, state: RecyclerView.State?, position: Int) {
+        val smoothScroller = TopSnappedSmoothScroller(this, recyclerView.context)
+        smoothScroller.targetPosition = position
+        startSmoothScroll(smoothScroller)
+    }
+}
+
+private class TopSnappedSmoothScroller(private val layoutManager: SnappingLinearLayoutManager, context: Context) : LinearSmoothScroller(context) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: Does `layoutManager` need to be a `SnappingLinearLayoutManager`? This doesn't seem to use anything specific that isn't in `LayoutManager`, so I'd apply LSP

**Review by rock3r — CHANGES_REQUESTED** [review]
Good start, but I think we need to smooth out a couple of things and check for a potential bug I fear may be lurking in there

**Comment by rock3r** [comment]
As discussed with the team, we could show the current slot until we're, say, 50% to 60% of the way through the slot (assuming that the slot is 75% talk and 25% overhead to change room etc), then switch to the next slot.

One note: our data model has no "slot" concept, so even if we don't change this, if the first event we happen to match on with the current time is not a "regular" slot (e.g., a 2h workshop), we may "skip" one or more slots. We should take that into account when looking for slots: we should get all the matching slots based on the start time (latest talks starting before now), and do calculations on the end/switch time based on the end time/duration of the shortest slot(s)

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt (hunk 7 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: Why is that? This class is (so far) only used here, so to me it makes sense to have it close to the place where it's used. If we find more uses for it (or it grows in terms of LOC) we can think about moving it to its separate file.

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 4 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: True. Do we have a timezone for the whole conference? Otherwise, we can assume that the timezone of any event will be the timezone for the conference, and we can use that to calculate the days accurately.

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 6 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Is there a data layer? 😄 Seriously though, were would you have it exactly?

Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt:38]: I'll have a look into it.

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 1 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: RE: the naming, you're right, it's confusing.

RE: the typed array, it was a way to try and enforce the fact that we should have only two elements in that collection. But if that's confusing/irrelevant, we can stick to a list.

Comment (rock3r) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Well, you only have `n` elements. You have two if the event is two days.

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 4 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Yes it's retrieved from Firebase. You can see how in `ScheduleService`

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt (hunk 7 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: Mostly in terms of mental encapsulation. It's a nit for me too, anyway, but if it's a de facto internal class maybe it should be an internal class then?

Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: Sure, it can be an internal class, but it must not be marked as `inner`. I placed it outside so to declutter the main class, but we can compromise on having it inside :)

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 9]
```diff
@@ -120,25 +124,51 @@ class SchedulePageView @JvmOverloads constructor(
...
+    }
+
+    private fun findNextEventForPage(page: SchedulePage) =
+        page
+            .events
+            .firstOrNull {
+                it.startTime.toDateTime(it.timeZone).isAfter(currentTime.currentLocalDateTime().toDateTime(it.timeZone))
+            }
+
+    private inner class ScrollingOnTabSelectedListener constructor(
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Could simply be `private inner class ScrollingOnTabSelectedListener(`

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt [hunk 10]
```diff
@@ -52,3 +57,19 @@ class ScheduleDayPageView @JvmOverloads constructor(
         override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) = oldEvents[oldItemPosition] == newEvents[newItemPosition]
     }
 }
+
+private class SnappingLinearLayoutManager(context: Context) : LinearLayoutManager(context) {
+
+    override fun smoothScrollToPosition(recyclerView: RecyclerView, state: RecyclerView.State?, position: Int) {
+        val smoothScroller = TopSnappedSmoothScroller(this, recyclerView.context)
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: I would avoid completely the `TopSnappedSmoothScroller ` and just go for 
```
val smoothScroller = object : LinearSmoothScroller(context) {
            override fun getVerticalSnapPreference() = SNAP_TO_START
}
```

File: app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt [hunk 11]
```diff
@@ -52,3 +57,19 @@ class ScheduleDayPageView @JvmOverloads constructor(
...
+        smoothScroller.targetPosition = position
+        startSmoothScroll(smoothScroller)
+    }
+}
+
+private class TopSnappedSmoothScroller(private val layoutManager: SnappingLinearLayoutManager, context: Context) : LinearSmoothScroller(context) {
+    override fun computeScrollVectorForPosition(targetPosition: Int): PointF? =
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/view/ScheduleDayPageView.kt]: We don't really need this, every built-in LayoutManager in RecyclerView implements `ScrollVectorProvider`, so the SmoothScroller will try to dispatch the call to the LayoutManager attached to it, no need to override it

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 12]
```diff
@@ -120,25 +124,51 @@ class SchedulePageView @JvmOverloads constructor(
...
+            private val schedule: Schedule,
+            private val viewPagerAdapter: ScheduleViewPagerAdapter
+    ) : OnTabSelectedListener {
+        override fun onTabReselected(tab: TabLayout.Tab) {
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Can we have an empty line after the class declaration? Don't know if ktlint caught this, but it's something we have been doing as a project convention

**Review by quiro91 — CHANGES_REQUESTED** [review]
Good work @danybony and @dextorer, I've just added a few comments to @rock3r 's review

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 13]
```diff
@@ -143,32 +143,37 @@ class SchedulePageView @JvmOverloads constructor(
     private fun findNextEventForPage(page: SchedulePage) =
         page
             .events
-            .firstOrNull {
-                it.startTime.toDateTime(it.timeZone).isAfter(currentTime.currentLocalDateTime().toDateTime(it.timeZone))
+            .firstOrNull { event ->
+                val startDateTime = event.startTime.toDateTime(event.timeZone)
+                val currentDateTime = currentTime.currentLocalDateTime().toDateTime(event.timeZone)
+                startDateTime.isAfter(currentDateTime)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: 💯

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 14]
```diff
@@ -120,25 +124,56 @@ class SchedulePageView @JvmOverloads constructor(
...
+
+        tabstrip.addOnTabSelectedListener(ScrollingOnTabSelectedListener(schedule, viewPagerAdapter))
         progressbar.visibility = View.GONE
     }
 
-    private class TrackingOnTabSelectedListener constructor(
-            private val analytics: Analytics,
+    private fun findTodayIndexOrDefault(pages: List<SchedulePage>): Int {
+        val now = currentTime.currentLocalDateTime()
+        return pages.firstOrNull { it.date.toLocalDate().isEqual(now.toLocalDate()) }
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: instead of doing 2 cycles (`firstOrNull` and then `indexOf`), you can use `indexOfFirst`with this predicate, then defaulting to 0 if the result is -1

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/index-of-first.html

**Review by fourlastor — COMMENTED** [review]
one minor comment

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 14 above)
Comment (fourlastor) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: possible clean implementation can be:

```kotlin
return pages.indexOfFirst { ... }.let {
  when(it) {
    -1 -> 0
    else -> it
  }
}
```

File: app/src/main/java/net/squanchy/support/system/CurrentTime.kt (hunk 2 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: Do you have a nice way of doing that? If you use the debug flag, the "debug" value must be written somewhere in the codebase. Otherwise we can create a debug version (as in, use the debug flavor) of `CurrentTimeModule` and provide a different value.

File: app/src/main/java/net/squanchy/support/system/CurrentTime.kt (hunk 3 above)
Comment (fourlastor) [reply app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: Actually right now you don't need to do anything to modules, extracting an interface and a different implementation to pass in unit tests will suffice

Comment (rock3r) [reply app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: Yeah, I was thinking about having a `DebugApplicationComponent` where you could inject this but yeah it should only be necessary in unit tests anyway, so can be easier.

Comment (dextorer) [reply app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: What I've done yesterday night (not tested though) is extracting an interface that is implemented by a class per each flavor (debug/release). Then I do the switch in `CurrentTimeModule`, like so:

```
@Module
class CurrentTimeModule {
    @Provides fun provideCurrentTime(): CurrentTime = 
        if (BuildConfig.DEBUG)
            DebugCurrentTime() 
        else 
            AndroidCurrentTime()
}
```

We can control the behavior of `DebugCurrentTime` as we please, leaving `AndroidCurrentTime` untouched.

Having a mock/debug time is not only useful for unit tests, but also when developing features. So I'd rather have a safe and easy way of doing that, and I think this is a good compromise. Opinions?

Comment (rock3r) [reply app/src/main/java/net/squanchy/support/system/CurrentTime.kt]: I'm ok with that for now :+1:

**Comment by fourlastor** [comment]
Only one comment: you're creating the current time inside lambdas in various places, can you create it once outside of it?

File: app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt [hunk 15]
```diff
@@ -1,12 +1,37 @@
...
+
+    fun findTodayIndexOrDefault(currentTime: CurrentTime) =
+        pages
+            .indexOfFirst { page ->
+                val now = currentTime.currentDateTime().toDateTime(timezone)
+                page.date.toLocalDate().isEqual(now.toLocalDate())
+            }
+            .let {
+                when (it) {
+                    -1 -> 0 // default to the first page
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt]: Nit: extract `-1` as `NOT_FOUND_INDEX` and `0` to `FIRST_PAGE_INDEX` so you can remove the comment

**Review by rock3r — APPROVED** [review]
Nice one @dextorer! Leaving to @quiro91 and @fourlastor — I've only left one tiny nitpicky comment :)

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt [hunk 16]
```diff
@@ -120,25 +123,41 @@ class SchedulePageView @JvmOverloads constructor(
...
-    private class TrackingOnTabSelectedListener constructor(
-            private val analytics: Analytics,
+    private interface OnTabSelectedListener : TabLayout.OnTabSelectedListener {
+
+        override fun onTabReselected(tab: TabLayout.Tab) {}
+        override fun onTabUnselected(tab: TabLayout.Tab) {}
+        override fun onTabSelected(tab: TabLayout.Tab) {}
+    }
+
+    private inner class ScrollingOnTabSelectedListener(
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: I might be missing something but is there any reason why the inner qualifier is required?

File: app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt [hunk 17]
```diff
@@ -1,12 +1,37 @@
...
+                    else -> it
+                }
+            }
+
+    fun findNextEventForPage(page: SchedulePage, currentTime: CurrentTime) =
+        page
+            .events
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt]: we can inline this event with page like
```
page.events
    .everythingElse()
```

**Review by quiro91 — CHANGES_REQUESTED** [review]
💯 
Some minor stuff left and a question but overall looks great

**Review by quiro91 — APPROVED** [review]

File: app/src/main/java/net/squanchy/schedule/SchedulePageView.kt (hunk 16 above)
Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/SchedulePageView.kt]: Because `currentTime` currently lives inside the outer class. We can pass it it as a dependency!

**Comment by dextorer** [comment]
@fourlastor Can you be more specific? You mean inside `Schedule`?

File: app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt [hunk 18]
```diff
@@ -1,12 +1,50 @@
...
     companion object {
 
-        fun create(pages: List<SchedulePage>) = Schedule(pages)
+        fun create(pages: List<SchedulePage>, timezone: DateTimeZone) = Schedule(pages, timezone)
     }
+
+    fun findTodayIndexOrDefault(currentTime: CurrentTime) =
+        pages
+            .indexOfFirst { page ->
+                val now = currentTime.currentDateTime().toDateTime(timezone)
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt]: here, extracting outside of the lambda (inside the method) so you don't create a date per page

File: app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt [hunk 19]
```diff
@@ -1,12 +1,50 @@
...
+                    NOT_FOUND_INDEX -> FIRST_PAGE_INDEX
+                    else -> it
+                }
+            }
+
+    fun findNextEventForPage(page: SchedulePage, currentTime: CurrentTime) =
+        page.events
+            .firstOrNull { event ->
+                val startDateTime = event.startTime.toDateTime().withZone(event.timeZone)
+                val currentDateTime = currentTime.currentDateTime().toDateTime().withZone(event.timeZone)
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt:39]: and here, extracting the current time inside the method but outside the lambda

Comment (dextorer) [reply app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt:39]: I can't really do it here, since we need the timezone for the current event. We could assume equal timezones across events, but that conflicts with the data model (having a timezone as part of each `Event`).

Comment (fourlastor) [reply app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt:39]: ah yeah makes sense
